### ISSUE TITLE
Compatibility with Keras Model?

### ISSUE THREAD

**Issue by ahundt** [ppwwyyxx/tensorpack#160]
Can I use a keras model to generate my TF layers and then train with tensorpack, or does tensorpack define attributes that would be incompatible?

**Comment by ppwwyyxx** [comment]
Theoretically you can use any symbolic functions in tensorpack. This includes Keras layers as long as you can call it on a tensorflow tensor,  e.g.:
```python
from keras.layers import Dense
...
logits = Dense(10, activation=None)(logits)    # logits is a tf.Tensor
```

But there are some issues:
1. If you use any keras layers **with variables** in tensorpack, non of the inference functionality will work anymore (including validation on test set every epoch) . Because keras doesn't respect `tf.variable_scope`. It use `tf.Variable` to create variables instead of `tf.get_variable`. MultiGPU won't work either because of the same reason.
2. If you use any keras layers which has different training/testing behavior (e.g. dropout), keras require an extra "is_training" tensor to be fed into the graph. You'll need to add a callback:
```python
class KerasHook(Callback):
    def _before_run(self, ctx):
        return tf.train.SessionRunArgs(fetches=[],feed_dict={'keras_learning_phase:0':1})
```

**Comment by ahundt** [comment]
Just getting back to this now, I'm actually looking at multi-gpu training with segmentation. I can also run on a slurm cluster where each node has multiple gpus.

**Comment by ahundt** [comment]
There is a [TensorLayer example with Keras integration](https://github.com/zsdonghao/tensorlayer/blob/master/example/tutorial_keras.py), also created when I asked the same question there, perhaps that might be helpful?

**Comment by ppwwyyxx** [comment]
That example doesn't use any multi tower training or inference. It will also work here if you don't use multi tower training or inference.

**Comment by ppwwyyxx** [comment]
The above commit added a keras example.
Now people can define keras model and utilize the more efficient tensorpack trainers, and even multi gpu training!

Since I'm apparently not a keras user, I only tested a very simple keras model. Not sure how it goes with more complex models. Feel free to report errors you may encounter.

**Comment by ahundt** [comment]
I took a look and it seems some additional work may be required to save out the weights in Keras format?

Also do you think there may be a problem if Keras supplies the logits portion as well?

**Comment by ppwwyyxx** [comment]
We use standard TF checkpoint, but Keras has its own format if I recall it correctly, so definitely would need extra effort.

Tensorpack doesn't assume you're doing classification, or assume you want to print the error rate. So you need to write them with symbolic functions. Using keras functions would run as well but be less informative.

**Comment by ppwwyyxx** [comment]
Closing because the current status (a [mnist-keras example](https://github.com/ppwwyyxx/tensorpack/blob/master/examples/mnist-keras.py)) is probably all we can support. The use of `tf.Variable` everywhere in Keras makes it very hard to work with any other frameworks.

**Comment by ahundt** [comment]
@ppwwyyxx is a change that would make it easier to integrate with fairly straightforward?

**Comment by ppwwyyxx** [comment]
The fundamental reason is that calling `tf.Variable` will ignore variable scope reuse, which is the fundamental mechanism of multigpu training. I'm not sure is this necessary for Keras or is it possible to change. There is actually a question on Keras about this: https://github.com/fchollet/keras/issues/7992

The current way to build keras model inside tensorpack isn't too bad IMHO. Apart from that maybe more things can be done on tensorpack side, e.g. integrate with Keras loss/regularization automatically.

**Comment by ppwwyyxx** [comment]
https://github.com/ppwwyyxx/tensorpack/blob/c2c895abcf8e7e13bc8cca9f1cee4a8631b9d341/examples/mnist-keras-v2.py#L40-L71

A WIP to train keras model easier..

**Comment by ppwwyyxx** [comment]
https://github.com/ppwwyyxx/tensorpack/blob/130f60ac4491d55a13632eabbad839c2e23138ba/examples/mnist-keras-v2.py#L35-L59

A more keras-like API which runs faster than keras.
It supports multi-gpu, but there are potential issues for now, e.g. weight-decay is probably not performed correctly.

**Comment by ahundt** [comment]
Interesting! Can networks like densenet and resnet can be constructed? These rely on the functional API.

**Comment by ppwwyyxx** [comment]
Presumably it should be able to train any Keras model. Although there might be unknown issues because I haven't tried anything other than mnist yet.

**Comment by ppwwyyxx** [comment]
Played with this a little bit yesterday. This experimental interface is too Keras-like and therefore cannot be made efficient to the best of my knowledge. Both of the two best-known scaling techniques (replicated and parameter_server mode in https://github.com/tensorflow/benchmarks) on single-machine require the use of variable scope. Therefore I cannot let users create their Keras models directly under a root variable scope -- it can never reaches the best performance. Models have to be created by tensorpack like the way the old [mnist-keras example](https://github.com/ppwwyyxx/tensorpack/blob/master/examples/mnist-keras.py) is doing. But the rest of the interface can probably still be Keras-like.

**Comment by ArtificialNotImbecile** [comment]
I can train a model using some keras layers by mimic the example code, but the saved checkpoint is problematic.  
I defined a model like this:
```
# some layer definition
@memoized
def keras_rnn():
    with clear_tower0_name_scope():
        blstm = layers.Bidirectional(
            layers.CuDNNLSTM(HIDDEN, return_state=True, return_sequences=True))
    return blstm

@memoized
def keras_rnn2():
    with clear_tower0_name_scope():
        blstm = layers.Bidirectional(
            layers.CuDNNLSTM(HIDDEN, return_state=False, return_sequences=True))
    return blstm

feat = ...  # extract features using tf.nn standard layers
blstm1 = keras_rnn()
blstm2 = keras_rnn2()
outputs1, *states1 = blstm1(feat)#stacked lstm layer1
outputs2 = blstm2(outputs1, initial_state=states1) #stacked lstm layer2
if ctx.is_main_training_tower:
    for op in blstm1.updates:
        tf.add_to_collection(tf.GraphKeys.UPDATE_OPS, op)
    for op in blstm2.updates:
        tf.add_to_collection(tf.GraphKeys.UPDATE_OPS, op)

...

```
After training and export model using:
```
predcfg = PredictConfig(
            model = MyModel(),
            session_init=SmartInit('/train_log/crnn_stacked/model-10000'),
            input_names=['image'],
            output_names=['decoded_value', 'decoded_indice', 'decoded_shape'])
ModelExporter(predcfg).export_compact('./crnn_keras.pb', optimize=False)
```
which raise an error caused by mismatch of checkpoint and model:  
>[1215 10:22:52 @sessinit.py:87] WRN The following variables are in the graph, but not found in the checkpoint: bidirectional/forward_cu_dnnlstm/kernel, bidirectional/forward_cu_dnnlstm/recurrent_kernel, bidirectional/forward_cu_dnnlstm/bias, bidirectional/backward_cu_dnnlstm/kernel, bidirectional/backward_cu_dnnlstm/recurrent_kernel, bidirectional/backward_cu_dnnlstm/bias, bidirectional_1/forward_cu_dnnlstm_1/kernel, bidirectional_1/forward_cu_dnnlstm_1/recurrent_kernel, bidirectional_1/forward_cu_dnnlstm_1/bias, bidirectional_1/backward_cu_dnnlstm_1/kernel, bidirectional_1/backward_cu_dnnlstm_1/recurrent_kernel, bidirectional_1/backward_cu_dnnlstm_1/bias
...
>[1215 10:22:52 @sessinit.py:87] WRN The following variables are in the checkpoint, but not found in the graph: beta1_power, beta2_power, global_step, learning_rate, tower0/bidirectional/backward_cu_dnnlstm/bias, tower0/bidirectional/backward_cu_dnnlstm/kernel, tower0/bidirectional/backward_cu_dnnlstm/recurrent_kernel, tower0/bidirectional/forward_cu_dnnlstm/bias, tower0/bidirectional/forward_cu_dnnlstm/kernel, tower0/bidirectional/forward_cu_dnnlstm/recurrent_kernel, tower0/bidirectional_1/backward_cu_dnnlstm_1/bias, tower0/bidirectional_1/backward_cu_dnnlstm_1/kernel, tower0/bidirectional_1/backward_cu_dnnlstm_1/recurrent_kernel, tower0/bidirectional_1/forward_cu_dnnlstm_1/bias, tower0/bidirectional_1/forward_cu_dnnlstm_1/kernel, tower0/bidirectional_1/forward_cu_dnnlstm_1/recurrent_kernel
...
>Failed precondition: Error while reading resource variable bidirectional_1/backward_cu_dnnlstm_1/kernel from Container: localhost. This could mean that the variable was uninitialized
